main/src/ca/uwaterloo/flix/language/phase/optimizer/Inliner.scala

```diff
@@ -77,8 +78,317 @@ object Inliner {
+    * When a variable is visited, it replaces the stale variable with the fresh one.
+    * Top-level function parameters are not substituted unless inlined, in which case
+    * the parameters are let-bound and added to the variable substitution.
+    */
+  private def visitExp(exp0: Expr, ctx0: LocalContext)(implicit sym0: Symbol.DefnSym, root: OccurrenceAst.Root, flix: Flix): Expr = exp0 match {
+    case Expr.Cst(cst, tpe, loc) =>
+      Expr.Cst(cst, tpe, loc)
+
+    case Expr.Var(sym, tpe, loc) =>
+      // Check for renamed local binder
+      ctx0.varSubst.get(sym) match {
+        case None => // Function parameter occurrence
+          Expr.Var(sym, tpe, loc)
+
+        case Some(freshVarSym) =>
+          ctx0.subst.get(freshVarSym) match {
+            // Case 1:
+            // The variable `sym` is in the substitution map. Replace `sym` with `e1`.
+            case Some(e1) =>
+              e1 match {
+                case SubstRange.DoneExpr(e) => // Reduced expression
+                  e
+
+                case SubstRange.SuspendedExpr(exp) => // Reduce suspended expr
+                  visitExp(exp, ctx0)
+              }
+
+            // Case 2:
+            // The variable `sym` is not in the substitution map, but is considered for inlining.
+            case None =>
+              ctx0.inScopeVars.get(freshVarSym) match {
+                case Some(_) => Expr.Var(freshVarSym, tpe, loc)
+                case None => Expr.Var(freshVarSym, tpe, loc)
+              }
+          }
+      }
+
+    case Expr.Lambda(fparam, exp, tpe, loc) =>
+      val (fp, varSubst1) = freshFormalParam(fparam)
+      val varSubst2 = ctx0.varSubst ++ varSubst1
+      val inScopeVars1 = ctx0.inScopeVars + (fp.sym -> BoundKind.ParameterOrPattern)
+      val ctx = ctx0.copy(varSubst = varSubst2, inScopeVars = inScopeVars1)
+      val e = visitExp(exp, ctx)
+      Expr.Lambda(fp, e, tpe, loc)
+
+    case Expr.ApplyAtomic(op, exps, tpe, eff, loc) =>
+      val es = exps.map(visitExp(_, ctx0))
+      Expr.ApplyAtomic(op, es, tpe, eff, loc)
+
+    case Expr.ApplyClo(exp1, exp2, tpe, eff, loc) =>
+      val exprCtx = ExprContext.AppCtx(exp2, ctx0.subst, ctx0.exprCtx)
+      val ctx = ctx0.copy(exprCtx = exprCtx)
+      val e1 = visitExp(exp1, ctx)
+      val e2 = visitExp(exp2, ctx0)
+      Expr.ApplyClo(e1, e2, tpe, eff, loc)
+
+    case Expr.ApplyDef(sym, exps, itpe, tpe, eff, loc) =>
+      val es = exps.map(visitExp(_, ctx0))
+      Expr.ApplyDef(sym, es, itpe, tpe, eff, loc)
+
+    case Expr.ApplyLocalDef(sym, exps, tpe, eff, loc) =>
+      ctx0.varSubst.get(sym) match {
+        case Some(freshVarSym) =>
+          val es = exps.map(visitExp(_, ctx0))
+          Expr.ApplyLocalDef(freshVarSym, es, tpe, eff, loc)
+
+        case None =>
+          throw InternalCompilerException("unexpected stale local def symbol", loc)
+      }
+
+    case Expr.Let(sym, exp1, exp2, tpe, eff, occur, loc) =>
+      if (isDead(occur)) {
+        if (isPure(exp1.eff)) {
```

Comment:
The if-stuff comes from the old code. But maybe a pattern match would be more appropriate?

```
case (Dead, Pure) =>
case (Dead, _) => ...
case (Once, Pure) => 
```

Why not?

Reply:
Sure, sounds good